### PR TITLE
etcd3/watcher: cancelling context shouldn't return error

### LINKED ISSUE #24528: unit flake: k8s.io/kubernetes/pkg/storage/etcd3
It failed in https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/24291/kubernetes-pull-test-unit-integration/22014/

```
--- FAIL: TestWatch (0.81s)
	watcher_test.go:204: #1: ResultChan should have been closed
```

cc @wojtek-t @xiang90

### PR DESCRIPTION
**Pull request by hongchaodeng**
Fixes #24528

### CODE REVIEW

**Comment by hongchaodeng** [comment]
This is blocking other people running PR tests. Please prioritize it.

**Comment by xiang90** [comment]
LGTM

**Comment by hongchaodeng** [comment]
Updated the test in the PR.
The reason is to reproduce the failure without the fix:
- The previous change is probabilistic to fail.
- Well, actually the new test is also probabilistic by itself, but more probable.
The probability happens in
```Go
	select {
	case wc.errChan <- err:
	case <-wc.ctx.Done():
	}
```
But I can repro the failure more easily (without the fix).

**Comment by xiang90** [comment]
@hongchaodeng Can we make the test itself reliable?

**Comment by hongchaodeng** [comment]
Seems like context.Canceled isn't the expected error. gRPC wraps it:

https://github.com/grpc/grpc-go/blob/61e92eacc3bc5ec3b227ed739d691cfb5034b8a4/rpc_util.go#L319-L327

**Comment by xiang90** [comment]
@hongchaodeng https://github.com/grpc/grpc-go/blob/61e92eacc3bc5ec3b227ed739d691cfb5034b8a4/codes/codes.go#L48

**Comment by hongchaodeng** [comment]
@xiang90 
We can't get the code unless we parse the string, which is an internal impl.

File: pkg/storage/etcd3/watcher.go [hunk 1]
```diff
@@ -276,6 +281,11 @@ func parseError(err error) *watch.Event {
 }
 
 func (wc *watchChan) sendError(err error) {
+	// Context.canceled is an expected behavior.
+	// We should just stop all goroutines in watchChan without returning error.
+	if err == context.Canceled {
```
Comment (xiang90) [inline pkg/storage/etcd3/watcher.go]: probably 
1. type assert on grpc error and then compare the code. 
2. if not grpc, compare with context.Canceled.

I think this is kind of etcd/clientv3 issue. We should not leak gRPC error. But this is P1 for k8s, so let's fix this in k8s.

/cc @heyitsanthony

**Comment by hongchaodeng** [comment]
Ah.. I think we can use `grpc.Code(err)`

File: pkg/storage/etcd3/watcher.go [hunk 2]
```diff
@@ -276,6 +283,12 @@ func parseError(err error) *watch.Event {
 }
 
 func (wc *watchChan) sendError(err error) {
+	// Context.canceled is an expected behavior.
+	// We should just stop all goroutines in watchChan without returning error.
+	// TODO: etcd client should return context.Canceled instead of grpc specific error.
+	if grpc.Code(err) == codes.Canceled {
```
Comment (xiang90) [inline pkg/storage/etcd3/watcher.go]: if grpc.Code(err) == codes.Canceled || err == context.Canceled

File: pkg/storage/etcd3/watcher.go [hunk 3]
```diff
@@ -276,6 +283,12 @@ func parseError(err error) *watch.Event {
 }
 
 func (wc *watchChan) sendError(err error) {
+	// Context.canceled is an expected behavior.
+	// We should just stop all goroutines in watchChan without returning error.
+	// TODO: etcd client should return context.Canceled instead of grpc specific error.
+	if grpc.Code(err) == codes.Canceled || err == context.Canceled {
```
Comment (wojtek-t) [inline pkg/storage/etcd3/watcher.go]: Why do we need grpc dependency?
We are still not using grpc anywhere in kubernetes.

**Comment by hongchaodeng** [comment]
@wojtek-t @gmarek 
Can you help take a look and re-apply 'lgtm'? Also re-trigger the Travis :P

File: pkg/storage/etcd3/watcher.go (hunk 3 above)
Comment (hongchaodeng) [inline pkg/storage/etcd3/watcher.go]: See [comment](https://github.com/kubernetes/kubernetes/pull/24638#discussion_r60678991).

Currently etcd client returns grpc error. It shouldn't though. We will fix it later by return "context.Canceled"

Comment (wojtek-t) [inline pkg/storage/etcd3/watcher.go]: The comment explains the problem (which I understand) not why you are using grpc.

>
Currently etcd client returns grpc error. It shouldn't though. We will fix it later by return "context.Canceled"

ok - that makes sense.

**Comment by wojtek-t** [comment]
LGTM

Travis seems to be broken today. I'm not sure we can do much about it.

**Pull request merged**